File: src/algorithms/arrays/sliding-window/min-size-subarray-sum/__tests__/min-size-subarray-sum.test.ts
import { describe, it, expect } from "vitest";
import { minSizeSubarraySum } from "../sources/min-size-subarray-sum.ts?fn";

describe("minSizeSubarraySum", () => {
  it("finds the shortest window in the default input", () => {
    /* [2,3,1,2,4,3], target=7: shortest is [4,3] = length 2 at index 4 */
    const result = minSizeSubarraySum([2, 3, 1, 2, 4, 3], 7);
    expect(result.minLength).toBe(2);
    expect(result.startIndex).toBe(4);
  });

  it("returns length 1 when a single element meets the target", () => {
    const result = minSizeSubarraySum([1, 4, 4], 4);
    expect(result.minLength).toBe(1);
  });

  it("returns 0 when no subarray sums to target", () => {
    const result = minSizeSubarraySum([1, 1, 1, 1], 10);
    expect(result.minLength).toBe(0);
  });

  it("handles the entire array being the minimum window", () => {
    const result = minSizeSubarraySum([1, 2, 3], 6);
    expect(result.minLength).toBe(3);
    expect(result.startIndex).toBe(0);
  });

  it("handles empty array", () => {
    const result = minSizeSubarraySum([], 7);
    expect(result.minLength).toBe(0);
  });

  it("handles target of zero or negative", () => {
    const result = minSizeSubarraySum([1, 2, 3], 0);
    expect(result.minLength).toBe(0);
  });

  it("handles single element equal to target", () => {
    const result = minSizeSubarraySum([7], 7);
    expect(result.minLength).toBe(1);
    expect(result.startIndex).toBe(0);
  });

  it("handles all elements the same", () => {
    /* [3,3,3,3], target=6: first window of length 2 satisfies */
    const result = minSizeSubarraySum([3, 3, 3, 3], 6);
    expect(result.minLength).toBe(2);
  });

  it("handles large values", () => {
    const result = minSizeSubarraySum([100, 1, 1, 1, 1], 100);
    expect(result.minLength).toBe(1);
    expect(result.startIndex).toBe(0);
  });
});

File: src/algorithms/arrays/sliding-window/min-size-subarray-sum/__tests__/min-size-subarray-sum_test.py
import importlib
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "sources"))
_mod = importlib.import_module("min-size-subarray-sum")
min_size_subarray_sum = _mod.min_size_subarray_sum

if __name__ == "__main__":
    result = min_size_subarray_sum([2, 3, 1, 2, 4, 3], 7)
    assert result["min_length"] == 2 and result["start_index"] == 4, f"Got {result}"

    result = min_size_subarray_sum([1, 4, 4], 4)
    assert result["min_length"] == 1, f"Got {result}"

    result = min_size_subarray_sum([1, 1, 1, 1], 10)
    assert result["min_length"] == 0, f"Got {result}"

    result = min_size_subarray_sum([1, 2, 3], 6)
    assert result["min_length"] == 3 and result["start_index"] == 0, f"Got {result}"

    result = min_size_subarray_sum([], 7)
    assert result["min_length"] == 0, f"Got {result}"

    result = min_size_subarray_sum([1, 2, 3], 0)
    assert result["min_length"] == 0, f"Got {result}"

    result = min_size_subarray_sum([7], 7)
    assert result["min_length"] == 1 and result["start_index"] == 0, f"Got {result}"

    result = min_size_subarray_sum([3, 3, 3, 3], 6)
    assert result["min_length"] == 2, f"Got {result}"

    result = min_size_subarray_sum([100, 1, 1, 1, 1], 100)
    assert result["min_length"] == 1 and result["start_index"] == 0, f"Got {result}"

    print("All tests passed!")
